### ISSUE TITLE
Prominently mention JWT/localStorage issues in FAQ -- Avoid JWT?

### ISSUE THREAD

**Issue by cmawhorter** [aws-samples/amazon-cognito-passwordless-auth#165]
I read [this answer in the FAQ](https://github.com/aws-samples/amazon-cognito-passwordless-auth/tree/main?tab=readme-ov-file#how-have-you-tested-the-security-posture-of-this-solution) and I was pretty far down the road before I finally did 2+2 and realized this [significant caveat regarding JWT/localStorage](https://github.com/aws-samples/amazon-cognito-passwordless-auth/blob/10c696371b84e41fcbcd8033f314bb6f73f79d1e/README.md#token-jwt-storage). Even being well aware of this pitfall with JWT I assumed it was a non-issue based on the FAQ. To help dummies like me in the future, I'd recommend adding that link to the amplify issue into the FAQ to drive the point home.

Totally my fault, but that leaves me at the point of either having to start over or keep moving forward and figure out a solution to avoid storing tokens in the browser.

I'm wondering: Has any thought/effort been given to using cloudfront instead of api gateway? I don't have a complete grasp of how everything is working here, but it seems like doing that would eliminate CORS entirely and make it pretty simple to support secure session cookies instead of JWT.

As it stands now, I think my plan is to hack together a lambda@edge proxy to the api gateway that stores tokens in DDB with an opaque session ID, and issues a secure cookie. It seems like that's my best worst option at this point, unless there is a better idea out there?

That'll mean forking the client side stuff, but that feels like a better option than abandoning all the super helpful server things in cdk.

**Comment by cmawhorter** [comment]
Understood. My intent with opening the issue was to suggest calling out the JWT requirement in the readme, and I'm definitely not looking to rehash the pro/con of jwt generally. 

Why cognito? Short answer is because identity and especially MFA is hard to get right and disastrous to get wrong.

Longer answer is that it has been many years (pre user pools) since I've looked at cognito, and I was giving it another shot. I'm essentially coming to cognito new again, and I was blindly assuming (hoping) things would be a lot more drop-in in the context of an app user auth scenario.

Re. state: 
I'm not sure I'm following how swapping in an opaque session id in place of jwt would change things. It just moves that sensitive state from the browser to the server, but is otherwise identical?

I'm guessing the underlying reason for leaving jwt here was because cognito/IAM can be used to auth* calls to aws resources? That is something I've stayed away from because it just feels like a bad idea, and I probably will never be convinced otherwise.

I donno. This all kinda took the wind out of my sails and I've moved on to other projects. I'll come back to the project at some point, but I don't know what I'm going to do.

**Comment by ottokruse** [comment]
Hi mate. Cognito can be a fine choice and worth your look again.

I think the issue you uncovered is, how do I store JWTs securely in the browser?

And the gist of it is, there is no definite answer on this, and "it depends". This is not particular for Cognito, all IDPs have this concern.

How can we get the wind back in your sails?

**Comment by cmawhorter** [comment]
> how do I store JWTs securely in the browser

We're wading into jwt debate territory, but I'm in the camp that _you can't_. Sure, from the context of "the browser" you can, but there is all sorts of malware out there that scoops up jwt (and session cookies), but jwt contain PII and are designed to be passed around and can more easily be leveraged by an attacker.

> [...] not particular for Cognito, all IDPs have this concern

True, but google, etc. have tons of resources and teams of people working to make sure things are safe and secure. Cognito offloads a lot of that responsibility onto me, which makes me ask... what's the point of cognito? If I have to give serious thought to all this then rolling my own is starting to look much more attractive as I'd have complete control and understanding of everything. This definitely feels like a least bad option situation.

> How can we get the wind back in your sails?

A lot of my issues are turning out to be with cognito and not exactly related to this repo, so I'm not sure you can. Everything here worked great and was easy to get running. Maybe pass up the AWS chain to stop shipping MVPs lol. I don't know.

**Comment by ottokruse** [comment]
> Everything here worked great and was easy to get running.

Thanks that's great to hear. 

If you can provide more details on your exact scenario I can think along with you better. I realize I've been shooting from the hip a bit in this discussion, without understanding your scenario well. 

I'd also be curious what features from the likes of Google you're referring to exactly, that you miss in Cognito in context of your current use case.

Otherwise, thank you for your interest in this solution here so far.

**Comment by cmawhorter** [comment]
My particular scenario is pretty straightforward: It's a dashboard web app with an email/password login. No social/oauth or anything like that. I'd like to stop using my own solution and switch to something else as part of a larger rewrite. I found this repo while researching MFA and it seemed like a good starting point.

> what features from the likes of Google you're referring to exactly

I was speaking within the context of google themselves consuming their jwt. The point I was trying to make was that me having to think about the security implications of using jwt is something I'd prefer not to do.

FWIW I came across [this](https://workos.com/user-management) randomly and it seems to be exactly what I'm after. A drop-in solution that has server session management, mfa, etc etc. I wasn't looking for it and didn't dive too deep beyond checking if it can do cookies instead of jwt, which it seems like it can (by storing tokens server side in db). I'd still prefer to use cognito though because I trust the inner workings of aws a lot more.
